services/audit/audit.go

```diff
@@ -0,0 +1,195 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package audit
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"time"
+
+	"code.gitea.io/gitea/models"
+	asymkey_model "code.gitea.io/gitea/models/asymkey"
+	auth_model "code.gitea.io/gitea/models/auth"
+	git_model "code.gitea.io/gitea/models/git"
+	organization_model "code.gitea.io/gitea/models/organization"
+	repository_model "code.gitea.io/gitea/models/repo"
+	secret_model "code.gitea.io/gitea/models/secret"
+	user_model "code.gitea.io/gitea/models/user"
+	webhook_model "code.gitea.io/gitea/models/webhook"
+	"code.gitea.io/gitea/modules/graceful"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/queue"
+	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/util/rotatingfilewriter"
+)
+
+type TypeDescriptor struct {
+	Type         string `json:"type"`
+	PrimaryKey   any    `json:"primary_key"`
+	FriendlyName string `json:"friendly_name"`
+	Target       any    `json:"-"`
+}
+
+type Event struct {
+	Action  Action         `json:"action"`
+	Doer    TypeDescriptor `json:"doer"`
+	Scope   TypeDescriptor `json:"scope"`
+	Target  TypeDescriptor `json:"target"`
+	Message string         `json:"message"`
+	Time    time.Time      `json:"time"`
```

Comment:
IP address?

Reply:
A lot of code logic which triggers audit events is implemented in services (that's a good thing). But there we usually don't have access to the incoming request and I don't think passing the ip address down the call chain is a good thing. For something like `doer` it's ok. We could maybe change `ctx.Doer` to a new type like
```go
type RequestUser struct {
	User
	RemoteAddress string
	...
}
```
Then we could pass such info around.

Reply:
Yes, source ip needs to be in context and is definitely very important in an audit log.

Reply:
Maybe we can have a middleware like `GetRealIP(ctx)` which can be invoked by services. I still think IP is necessary for audit log.

Reply:
I agree that IP address is pretty much required element for auditing

Reply:
I thought about something like this. Would be better to have a own type but without inheritance that's not really possible in Go:
```diff
diff --git a/models/user/user.go b/models/user/user.go
index 86cf2ad28..940fb9b00 100644
--- a/models/user/user.go
+++ b/models/user/user.go
@@ -8,6 +8,7 @@ import (
        "context"
        "encoding/hex"
        "fmt"
+       "net/http"
        "net/url"
        "path/filepath"
        "strings"
@@ -66,6 +67,11 @@ const (
        EmailNotificationsAndYourOwn = "andyourown"
 )
 
+type RequestInfo struct {
+       Request *http.Request
+       // UsedAuthMethod string
+}
+
 // User represents the object of individual and member of organization.
 type User struct {
        ID        int64  `xorm:"pk autoincr"`
@@ -140,6 +146,8 @@ type User struct {
        DiffViewStyle       string `xorm:"NOT NULL DEFAULT ''"`
        Theme               string `xorm:"NOT NULL DEFAULT ''"`
        KeepActivityPrivate bool   `xorm:"NOT NULL DEFAULT false"`
+
+       RequestInfo *RequestInfo `xorm:"-"`
 }
 
 func init() {
diff --git a/routers/api/v1/api.go b/routers/api/v1/api.go
index c71d0ae2d..45b5bacd6 100644
--- a/routers/api/v1/api.go
+++ b/routers/api/v1/api.go
@@ -120,6 +120,8 @@ func sudo() func(ctx *context.APIContext) {
                                        return
                                }
 
+                               user.RequestInfo = ctx.Doer.RequestInfo
+
                                audit.Record(audit.UserImpersonation, ctx.Doer, ctx.Doer, ctx.Doer, "User %s impersonating user %s.", ctx.Doer.Name, user.Name)
 
                                log.Trace("Sudo from (%s) to: %s", ctx.Doer.Name, user.Name)
diff --git a/services/auth/middleware.go b/services/auth/middleware.go
index 49e823c66..9b3e246e9 100644
--- a/services/auth/middleware.go
+++ b/services/auth/middleware.go
@@ -65,6 +65,10 @@ func authShared(ctx *context.Base, sessionStore SessionStore, authMethod Method)
                }
                ar.IsBasicAuth = ctx.Data["AuthedMethod"].(string) == BasicMethodName
 
+               ar.Doer.RequestInfo = &user_model.RequestInfo{
+                       Request: ctx.Req,
+               }
+
                ctx.Data["IsSigned"] = true
                ctx.Data[middleware.ContextDataKeySignedUser] = ar.Doer
                ctx.Data["SignedUserID"] = ar.Doer.ID
```
This could be expanded with a more detailed info how a user was authenticated (by basic auth, which token was used, ...)

I don't really like adding more stuff into `Context` because that's like the "god object" anti pattern.

Reply:
I would not recommend adding *http.Request pointer to doer request struct but probably only IP address

Reply:
Depends on what info we want to have. ip, user agent, ...